tools.rkt

```diff
@@ -305,6 +315,87 @@
                  (cons (cons i s) undo)))
           (else (invalid)))))
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Define explore state to maintain the data held at every step
+;; when exploring
+(struct explore-state (history choices results) #:prefab)
+(define (explore-state/can-undo? exp-state) (pair? (explore-state->history exp-state)))
+(define (explore-state/undo exp-state)
+  (cdar (explore-state->history exp-state)))
+(define (explore-state/make-choice exp-state choice)
+  (match (exp-state)
+    [(explore-state history _ _)
+     (let* ([expanded-stream (step (list-ref choices choice))]
+            [expanded-choices (stream->choices expnded-stream)]
+            [choices (dropf expanded-choices state?)]
+            [results (takef expanded-choices state?)])
+       ; NOTE: could store history as just a pointer to previous choice instead of list
+       (explore-state (cons exp-state history) choices results)]))
+(define (explore-state/make-choices exp-state choices)
+  (match choices
+    ['() exp-state]
+    [(cons c hoices)
+     (explore-state/make-choices (explore-state/make-choice exp-state c) hoices)]))
+
+(define (pp/results results qvars)
+  (unless (= (length results) 0)
+    (printf "Number of results: ~a\n" (length results))
+    (for-each (lambda (st)
+                (pp/qvars qvars (walked-term initial-var st))
+                (newline))
+              results)))
+(define (pp/choices choices qvars)
+  (if (= 0 (length choices))
+    (printf "No more choices available. Undo to continue.\n")
+    (printf "Number of Choices: ~a\n" (length choices))))
+    (for-each (lambda (i s)
+                (printf (string-append "\n" margin "Choice ~s:\n") (+ i 1))
+                (print-choice qvars s))
+              (range (length choices)) choices)
+(define (pp/explore-state exp-state qvars)
+  (match exp-state
+    [(explore-state history choices results)
+     (display "\n========================================")
+     (displayln "========================================")
+     (pp/results results qvars)
+     (when (and (previous-choice undo) (null? results))
+       (printf "Previous Choice:\n")
+       (print-choice qvars (previous-choice undo))
+       (newline))
+     (printf "Current Depth: ~a\n" (length undo))
+     (pp/choices choices qvars)
+     ]))
+
+(define (explore-state/stream step qvars strm)
```

Comment:
We also want to add another parameter for the interaction policy that determines how interaction I/O works.

For organizing your experiments, you really want to maintain multiple policies to see what impact each of them has on the agent's performance.  Being able to swap policies easily is important.

We can also redefine `explore/stream` later, having it call your new driver with a textual pretty-print-and-read interaction policy that mimics the current behavior.